### PR TITLE
[JENKINS-33068] AsyncPeriodicWork/AsyncAperiodicWork log files rotation

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-33068](https://issues.jenkins-ci.org/browse/JENKINS-33068)

Before this change the log file would be overwritten every execution... which makes finding any failures of a short period work almost impossible.

### CODE REVIEW

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 1]
```diff
@@ -40,8 +40,46 @@
...
-	/**
+    /**
+     * The default number of minutes after which to try and rotate the log file used by {@link #createListener()}.
+     * This value is controlled by the system property {@code hudson.model.AsyncAperiodicWork.logRotateMinutes}.
+     * Each individual AsyncAperiodicWork can also have a per-extension override using the system property
+     * based on their fully qualified class name with {@code .logRotateMinutes} appended.
+     *
+     * @since 1.650
+     */
+    private static Long LOG_ROTATE_MINUTES = Long.getLong(AsyncAperiodicWork.class.getName() + ".logRotateMinutes",
```
Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: final for good measure?

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 2]
```diff
@@ -40,8 +40,46 @@
...
+    /**
+     * The default file size after which to try and rotate the log file used by {@link #createListener()}.
+     * A value of {@code -1L} disables rotation based on file size.
+     * This value is controlled by the system property {@code hudson.model.AsyncAperiodicWork.logRotateSize}.
+     * Each individual AsyncAperiodicWork can also have a per-extension override using the system property
+     * based on their fully qualified class name with {@code .logRotateSize} appended.
+     *
+     * @since 1.650
+     */
+    private static Long LOG_ROTATE_SIZE = Long.getLong(AsyncAperiodicWork.class.getName() + ".logRotateSize", -1L);
```
Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: final for good measure?

Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: Why use a long and then "-1" for unset.  either use a `long` or use `null`?

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java (hunk 1 above)
Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: why a `Long` not a `long` - this has a default value so `null` seems invalid to me?

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 3]
```diff
@@ -59,19 +100,25 @@ protected AsyncAperiodicWork(String name) {
     public final void doAperiodicRun() {
         try {
             if(thread!=null && thread.isAlive()) {
-                logger.log(Level.INFO, name+" thread is still running. Execution aborted.");
+                logger.log(getSlowLoggingLevel(), name+" thread is still running. Execution aborted.");
```
Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: whilst you where here changing this line is should be paramaterized logging to avoid string concat if unneeded.

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 4]
```diff
@@ -59,19 +100,25 @@ protected AsyncAperiodicWork(String name) {
...
                 return;
             }
             thread = new Thread(new Runnable() {
                 public void run() {
-                    logger.log(Level.INFO, "Started "+name);
+                    logger.log(getNormalLoggingLevel(), "Started "+name);
```
Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: ditto String concat

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 5]
```diff
@@ -80,8 +127,8 @@ public void run() {
                         l.closeQuietly();
                     }
 
-                    logger.log(Level.INFO, "Finished "+name+". "+
-                        (System.currentTimeMillis()-startTime)+" ms");
+                    logger.log(getNormalLoggingLevel(), "Finished {0}. {1,number} ms",
+                            new Object[]{name, System.currentTimeMillis() - startTime});
```
Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: NIT: the time/duration in the log will be different to that in the tasklog.
(not a new issue)

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 6]
```diff
@@ -91,8 +138,39 @@ public void run() {
...
+                    File o = i == 0 ? f : new File(f.getParentFile(), f.getName() + "." + i);
+                    if (o.isFile()) {
+                        if (p != null && !p.exists()) {
+                            if (!o.renameTo(p)) {
+                                logger.log(getErrorLoggingLevel(), "Could not rotate log files {0} to {1}",
+                                        new Object[]{o, p});
+                            }
+                        } else {
+                            if (!o.delete()) {
+                                logger.log(getErrorLoggingLevel(), "Could not delete log files {0} to enable rotation",
```
Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: could not delete **log** file (singular)

File: core/src/main/java/hudson/util/StreamTaskListener.java [hunk 7]
```diff
@@ -98,6 +98,13 @@ public StreamTaskListener(File out, Charset charset) throws IOException {
         this(new FileOutputStream(out),charset);
     }
 
+    public StreamTaskListener(File out, boolean append, Charset charset) throws IOException {
```
Comment (jtnord) [inline core/src/main/java/hudson/util/StreamTaskListener.java]: javadoc missing for public ctor.

File: core/src/main/java/hudson/model/AsyncPeriodicWork.java [hunk 8]
```diff
@@ -84,7 +164,7 @@ protected StreamTaskListener createListener() {
      * Determines the log file that records the result of this task.
      */
     protected File getLogFile() {
-        return new File(Jenkins.getInstance().getRootDir(),name+".log");
+        return new File(Jenkins.getInstance().getRootDir(),"logs/tasks/"+name+".log");
```
Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncPeriodicWork.java]: +Long.MAX_LONG for not cluttering the JENKINS_HOME anymore!

**Comment by jtnord** [comment]
:bee: but some comments inline.

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java (hunk 5 above)
Comment (stephenc) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: yes but I had issues getting the stopped time and ensuring that the stop time would always be populated... I have thought some more and now have a code path that works

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 9]
```diff
@@ -40,8 +40,46 @@
...
+     * @since 1.650
+     */
+    private final long logRotateSize;
+    /**
+     * The last time the log files were rotated.
+     *
+     * @since 1.650
+     */
+    private Long lastRotateMillis;
+    /**
```
Comment (varmenise) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: 'long' for consistency ?

Comment (stephenc) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: I'm using the special value of `null` as a marker for first run... but we could use `0L` also...

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 10]
```diff
@@ -91,8 +138,39 @@ public void run() {
     }
 
     protected StreamTaskListener createListener() {
+        File f = getLogFile();
+        if (f.isFile()) {
+            if ((lastRotateMillis == null)
+                    || (lastRotateMillis + logRotateMillis < System.currentTimeMillis())
```
Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: lastRotateMillis will be null on first run so even if rotation is disabled it will be rotated once...

Is this intentional or a :big:

Comment (stephenc) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: intentional. clarified in 2316bfb

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 11]
```diff
@@ -91,8 +138,39 @@ public void run() {
...
+                                        new Object[]{o, p});
+                            }
+                        } else {
+                            if (!o.delete()) {
+                                logger.log(getErrorLoggingLevel(), "Could not delete log files {0} to enable rotation",
+                                        o);
+                            }
+                        }
+                        p = o;
+                    }
```
Comment (varmenise) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: so first iteration:
`i = 5` `o=fileName5` `p = null` --> `o` gets deleted, `p = fileName5` now and does not exist cause it was deleted

second iteration:
`i = 4` `o=fileName4` `p=fileName5` and `p` does not exist --> `o` gets renamed to `p` = `fileName5` then `p` is set to `o` which is `fileName5`

shouldn't be `p = fileName4` at this point?

File: core/src/main/java/hudson/model/AsyncPeriodicWork.java [hunk 12]
```diff
@@ -21,6 +22,45 @@
...
 public abstract class AsyncPeriodicWork extends PeriodicWork {
     /**
+     * The default number of minutes after which to try and rotate the log file used by {@link #createListener()}.
+     * This value is controlled by the system property {@code hudson.model.AsyncPeriodicWork.logRotateMinutes}.
+     * Each individual AsyncPeriodicWork can also have a per-extension override using the system property
+     * based on their fully qualified class name with {@code .logRotateMinutes} appended.
+     *
+     * @since 1.650
+     */
+    private static Long LOG_ROTATE_MINUTES = Long.getLong(AsyncPeriodicWork.class.getName() + ".logRotateMinutes",
```
Comment (varmenise) [inline core/src/main/java/hudson/model/AsyncPeriodicWork.java]: same as per @jtnord 's comment

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java (hunk 11 above)
Comment (stephenc) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: ahhh good catch

File: core/src/main/java/hudson/model/AsyncPeriodicWork.java [hunk 13]
```diff
@@ -21,6 +22,45 @@
...
+     * @since 1.650
+     */
+    private final long logRotateSize;
+    /**
+     * The last time the log files were rotated.
+     *
+     * @since 1.650
+     */
+    private Long lastRotateMillis;
+    /**
```
Comment (varmenise) [inline core/src/main/java/hudson/model/AsyncPeriodicWork.java]: same as per my comment above

**Comment by KostyaSha** [comment]
More system variables? I wish to see installation with `too many environment variables`

**Comment by stephenc** [comment]
@KostyaSha I do not see these as something that a user should ever need to configure... but just in case I leave the option to configure them... It would be nice if we had a config setup screen like firefox where we could register all of them globally to help discovery, but these are not ones I would anticipate exposing to regular admins... unless they are stuck with a fast rotating chatty periodic task that is overspilling their JENKINS_HOME every day... those rare cases will need the help of the system property

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 14]
```diff
@@ -91,8 +141,54 @@ public void run() {
...
     protected StreamTaskListener createListener() {
+        File f = getLogFile();
+        if (!f.getParentFile().isDirectory()) {
+            if (!f.getParentFile().mkdirs()) {
+                logger.log(getErrorLoggingLevel(), "Could not create directory {0}", f.getParentFile());
+            }
+        }
+        if (f.isFile()) {
+            if ((lastRotateMillis + logRotateMillis < System.currentTimeMillis())
+                    || (logRotateSize > 0 && f.length() > logRotateSize)) {
```
Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: no need to check `logRotateSize > 0`  just `f.length() > logRotateSize` now.

File: core/src/main/java/hudson/model/AsyncPeriodicWork.java [hunk 15]
```diff
@@ -73,8 +124,54 @@ public void run() {
...
+        } else {
+            lastRotateMillis = System.currentTimeMillis();
+            // migrate old log files the first time we start-up
+            File oldFile = new File(Jenkins.getActiveInstance().getRootDir(), f.getName());
+            if (oldFile.isFile()) {
+                File newFile = new File(f.getParentFile(), f.getName() + ".1");
+                if (!newFile.isFile()) {
+                    // if there has never been rotation then this is the first time
+                    if (oldFile.renameTo(newFile)) {
+                        logger.log(getNormalLoggingLevel(), "Moved {0} to {1}.1", new Object[]{oldFile, f});
```
Comment (jtnord) [inline core/src/main/java/hudson/model/AsyncPeriodicWork.java]: `newFile` and `{1}` rather than `f` and `{1}.1`so if you change the name - you do it in one place.

**Comment by jtnord** [comment]
NITs inline - but :bee:

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 16]
```diff
@@ -91,8 +141,54 @@ public void run() {
...
+                                        new Object[]{o, p});
+                            }
+                        } else {
+                            if (!o.delete()) {
+                                logger.log(getErrorLoggingLevel(), "Could not delete log file {0} to enable rotation",
+                                        o);
+                            }
+                        }
+                    }
+                    p = o;
```
Comment (rsandell) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: _nit_ the naming choice of these variables makes it a bit hard to follow along with what is supposed to happen.

**Comment by KostyaSha** [comment]
@stephenc i thought one of the previous PRs already added experimental features that temporary may host such configurations. Linux kernel has also many strange tunes, but it allow avoid restarts and filling system vars that not critical for jenkins container/process itself.

**Comment by varmenise** [comment]
:bee:

File: core/src/main/java/hudson/model/AsyncPeriodicWork.java [hunk 17]
```diff
@@ -73,8 +124,54 @@ public void run() {
...
 
     protected StreamTaskListener createListener() {
+        File f = getLogFile();
+        if (!f.getParentFile().isDirectory()) {
+            if (!f.getParentFile().mkdirs()) {
+                logger.log(getErrorLoggingLevel(), "Could not create directory {0}", f.getParentFile());
+            }
+        }
+        if (f.isFile()) {
+            if ((lastRotateMillis + logRotateMillis < System.currentTimeMillis())
```
Comment (rsandell) [inline core/src/main/java/hudson/model/AsyncPeriodicWork.java]: This looks very similar to what is implemented in `AsyncAperiodicWork` can it be moved to some common ancestor or a utility method somewhere for better reuse than having to know there are two places to fix in the future?

**Comment by stephenc** [comment]
@reviewbybees done ready for @jenkinsci/code-reviewers

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 18]
```diff
@@ -40,8 +40,48 @@
...
+     * @since 1.650
+     */
+    private static final long LOG_ROTATE_SIZE = Long.getLong(AsyncAperiodicWork.class.getName() + ".logRotateSize",
+            -1L);
+    /**
+     * The number of milliseconds (since startup or previous rotation) after which to try and rotate the log file.
+     *
+     * @since 1.650
+     */
+    private final long logRotateMillis;
```
Comment (rsandell) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: Shouldn't be final so that a system groovy script or similar can change the value in runtime.

File: core/src/main/java/hudson/model/AsyncPeriodicWork.java [hunk 19]
```diff
@@ -73,8 +124,54 @@ public void run() {
...
+                    // if there has never been rotation then this is the first time
+                    if (oldFile.renameTo(newFile)) {
+                        logger.log(getNormalLoggingLevel(), "Moved {0} to {1}.1", new Object[]{oldFile, f});
+                    } else {
+                        logger.log(getErrorLoggingLevel(), "Could not move {0} to {1}.1",
+                                new Object[]{oldFile, f});
+                    }
+                }
+            }
+        }
```
Comment (tfennelly) [inline core/src/main/java/hudson/model/AsyncPeriodicWork.java]: `AsyncPeriodicWork.createListener` and `AsyncAperiodicWork.createListener` look to be exactly the same. I don't mind a bit of duplication, but in this case (if I'm right and they are the same) I'd prefer to not see it because (as @rsandell pointed out) the code is not the easiest to read first time (nondescript variable names ... magic number `5` etc).

File: core/src/main/java/hudson/model/AsyncPeriodicWork.java (hunk 15 above)
Comment (KostyaSha) [inline core/src/main/java/hudson/model/AsyncPeriodicWork.java]: @jtnord didn't you already add normal logger into core?

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java (hunk 18 above)
Comment (stephenc) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: then you need synchronization handlers, etc.

File: core/src/main/java/hudson/model/AsyncPeriodicWork.java (hunk 17 above)
Comment (stephenc) [inline core/src/main/java/hudson/model/AsyncPeriodicWork.java]: There is no common ancestor

**Comment by tfennelly** [comment]
:bee:

File: core/src/main/java/hudson/model/AsyncPeriodicWork.java (hunk 17 above)
Comment (tfennelly) [inline core/src/main/java/hudson/model/AsyncPeriodicWork.java]: Can it be a static utility of some sort?

Comment (stephenc) [inline core/src/main/java/hudson/model/AsyncPeriodicWork.java]: Not in a nice way as it needs to store state in a field

File: core/src/main/java/hudson/model/AsyncAperiodicWork.java [hunk 20]
```diff
@@ -40,8 +40,48 @@
...
+    private static final long LOG_ROTATE_MINUTES = Long.getLong(AsyncAperiodicWork.class.getName() +
+            ".logRotateMinutes", TimeUnit.DAYS.toMinutes(1));
+    /**
+     * The default file size after which to try and rotate the log file used by {@link #createListener()}.
+     * A value of {@code -1L} disables rotation based on file size.
+     * This value is controlled by the system property {@code hudson.model.AsyncAperiodicWork.logRotateSize}.
+     * Each individual AsyncAperiodicWork can also have a per-extension override using the system property
+     * based on their fully qualified class name with {@code .logRotateSize} appended.
+     *
+     * @since 1.650
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: I know for a fact that this is wrong, as 1.650 is going to be just 1.649 + security fixes.

Comment (stephenc) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: Then why is master still towards 1.650-SNAPSHOT?

Comment (KostyaSha) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: Because nobody follows maven versioning rules in release management (mostly in plugins) :)
That's why the universal way would be to keep to `@since TODO`, though i not sure how such changes would be cherry-picked into LTS.

Comment (stephenc) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: @KostyaSha the point is that `master` is supposed to be the next version. If @daniel-beck as security lead has decreed that 1.650 shall be just 1.649 + security fixes, then the version number in `master` should have been adjusted accordingly.

The habit of using `@since TODO` or `@since FIXME` is a bad pattern as there are countless examples of them not actually being corrected for the merge... I now think it is much better to be explicit in the PR and complain if the PR is saying the wrong version against the branch that it will be merged into... which says that either @daniel-beck will need to fix up any `@since 1.650` versions when the 1.650 is being rolled... or @daniel-beck needs to bump the version on `master` to `1.651-SNAPSHOT` and then I need to fix my `@since` tags before merge

Comment (KostyaSha) [inline core/src/main/java/hudson/model/AsyncAperiodicWork.java]: > then the version number in master should have been adjusted accordingly.

That would lead to incorrect versions already committed in git's master. You will have versions mix in git history that is wrong from maven versioning viewpoint.

> is a bad pattern as there are countless examples of them not actually being corrected for the merge.

That are release management issues and not PR author. This should be adjusted either before release, either during merge (gatekeeper/validated). That's why i suggest not using 'merge' button at all for gatekeepering with github-pullrequest-plugin (only labels and comments for triggering /o/ merges where you can enforce additional checks).

**Comment by olivergondza** [comment]
Why can't we use `ReopenableRotatingFileOutputStream` for this (the same way slaves does)? I like how there is exactly one log file for one invocation. It would eliminate a lot of the constants as well.

**Comment by jtnord** [comment]
@olivergondza because when you do this every 5 minutes  (or even more often say) then you end up with a massive mess of files, or you the file that you are interested in has been rotated before you realize you need to look at it (problem happened over night for instance.

**Comment by daniel-beck** [comment]
@stephenc Could you document the new system properties on https://wiki.jenkins-ci.org/display/JENKINS/Features+controlled+by+system+properties?

**Comment by stephenc** [comment]
@daniel-beck done
